encoder/MysqlEncoder.php

```diff
@@ -0,0 +1,204 @@
+     * @param string $type SQL datatype type
+     *
+     * @return string a Maria COLUMN_GET expression
+     */
+    public function columnExpression($name, $type = 'char')
+    {
+        $modelClass = $this->modelClass;
+        $sql = '[[' . $modelClass::dynamicColumn() . ']]';
+        foreach (explode('.', $name) as $column) {
+            $sql = "COLUMN_GET($sql, '$column' AS $type)";
+        }
+
+        return $sql;
+    }
+
+    /**
+     * Creates a dynamic column SQL expression representing the given attributes.
+     *
+     * @param array $attrs the dynamic attributes, which may be nested
+     *
+     * @return null|\yii\db\Expression
+     */
+    public function dynColExpression($attrs) {
+        if (!$attrs) {
+            return null;
+        }
+
+        $params = [];
+
+        // todo For now we only have Maria. Add PgSQL and generic JSON.
+        static::encodeArrayForMaria($attrs);
+        $sql = static::dynColSqlMaria($attrs, $params);
+
+        return new \yii\db\Expression($sql, $params);
+    }
+
+    /**
+     * Decode a serialized blob of dynamic attributes.
+     *
+     * At present the only supported input format is JSON returned from Maria. It may work
+     * also for PostgreSQL.
+     *
+     * @param string $encoded Serialized array of attributes in DB-specific form
+     *
+     * @return array Dynamic attributes in name => value pairs (possibly nested)
+     */
+    public function dynColDecode($encoded)
+    {
+        // Maria has a bug in its COLUMN_JSON funcion in which it fails to escape the
+        // control characters U+0000 through U+001F. This causes JSON decoders to fail.
+        // This workaround escapes those characters.
+        $encoded = preg_replace_callback(
+            '/[\x00-\x1f]/',
+            function ($matches) {
+                return sprintf('\u00%02x', ord($matches[0]));
+            },
+            $encoded
+        );
+
+        $decoded = json_decode($encoded, true);
+        if ($decoded) {
+            static::decodeArrayForMaria($decoded);
+        }
+
+        return $decoded;
+    }
+
+    /**
+     * Encode as data URIs strings that JSON cannot express.
+     *
+     * @param mixed $value a value to encode
+     *
+     * @return string the encoded data URI
+     */
+    public static function encodeForMaria($value)
+    {
+        return is_string($value)
+        && (!mb_check_encoding($value, 'UTF-8') || strpos($value, self::DATA_URI_PREFIX) === 0)
+            ? self::DATA_URI_PREFIX . base64_encode($value)
+            : $value;
+    }
+
+    /**
+     * Decode strings encoded as data URIs.
+     *
+     * @param string $value the data URI to decode
+     *
+     * @return string the decoded value
+     */
+    public static function decodeForMaria($value)
+    {
+        return is_string($value) && strpos($value, self::DATA_URI_PREFIX) === 0
+            ? file_get_contents($value)
+            : $value;
+    }
```

Comment:
I think both `encodeForMaria()` and `decodeForMaria()` are badly named. Any JSON-oriented storage will need these methods. they can go in the base encoder and be renamed